drivers/metal/metal_objects.h

```diff
@@ -110,9 +81,61 @@ enum StageResourceUsage : uint32_t {
 	ComputeWrite = (MTLResourceUsageWrite << RDD::SHADER_STAGE_COMPUTE * 2),
 };
 
-typedef LocalVector<__unsafe_unretained id<MTLResource>> ResourceVector;
+typedef id<MTLResource> __unsafe_unretained MTLResourceUnsafe;
+
+template <>
+struct HashMapHasherDefaultImpl<MTLResourceUnsafe> {
+	static _FORCE_INLINE_ uint32_t hash(const MTLResourceUnsafe p_pointer) { return hash_one_uint64((uint64_t)p_pointer); }
+};
+
+typedef LocalVector<MTLResourceUnsafe> ResourceVector;
 typedef HashMap<StageResourceUsage, ResourceVector> ResourceUsageMap;
 
+struct ResourceUsageEntry {
+	StageResourceUsage usage = ResourceUnused;
+	uint32_t unused = 0;
+
+	ResourceUsageEntry() {}
+	ResourceUsageEntry(StageResourceUsage p_usage) :
+			usage(p_usage) {}
+};
+
+template <>
+struct is_zero_constructible<ResourceUsageEntry> : std::true_type {};
+
+/*! Track the cumulative usage for a resource during a render or compute pass */
+typedef HashMap<MTLResourceUnsafe, ResourceUsageEntry> ResourceToStageUsage;
+
+/*! Track resource and ensure they are resident prior to dispatch or draw commands.
+ *
+ * The primary purpose of this data structure is to track all the resources that must be made resident prior
+ * to issuing the next dispatch or draw command. It aggregates all resources used from argument buffers.
+ *
+ * As an optimization, this data structure also tracks previous usage for resources, so that
+ * it may avoid binding them again in later commands if the resource is already resident and its usage flagged.
+ */
+struct API_AVAILABLE(macos(11.0), ios(14.0), tvos(14.0)) ResourceTracker {
```

Comment:
Fixes GPU corruption / crashes by tracking resource usage and ensuring they are resident prior to each command (draw, dispatch, etc)